lib/find-variations/index.js

```diff
@@ -0,0 +1,728 @@
+    }
+
+    //run matching process using substrings and edit distance
+    if (useConsole) {
+      console.log(chalk.cyanBright.bold('\nMatching process starting...'));
+      startTime = new Date();
+    }
+
+    const getDDWikiUrlForResourceName = standardResourceName =>
+      getReferenceMetadata()?.resources?.find(item => item?.resourceName === standardResourceName)?.wikiPageURL ?? null;
+
+    Object.keys(metadataReportMap).forEach(resourceName => {
+      const isStandardResource = standardMetadataMap?.[resourceName] ?? false;
+
+      //check for resource variations if the resource name doesn't match the reference metadata exactly
+      if (!isStandardResource) {
+        Object.keys(standardMetadataMap).forEach(standardResourceName => {
+          const normalizedStandardResourceName = normalizeDataElementName(standardResourceName),
+            normalizedResourceName = normalizeDataElementName(resourceName),
+            isMinMatchingLength = resourceName?.length > MIN_MATCHING_LENGTH,
+            isSubstringMatch =
+              isMinMatchingLength &&
+              (normalizedResourceName.includes(normalizedStandardResourceName) ||
+                normalizedStandardResourceName.includes(normalizedResourceName)),
+            isExactMatch = normalizedResourceName === normalizedStandardResourceName && resourceName !== standardResourceName,
+            hasStandardResource = metadataReportMap?.[standardResourceName];
+
+          if (!hasStandardResource && (isExactMatch || isSubstringMatch)) {
+            const suggestion = {
+              resourceName,
+              suggestedResourceName: standardResourceName,
+              strategy: MATCHING_STRATEGIES.SUBSTRING,
+              ddWikiUrl: getDDWikiUrlForResourceName(standardResourceName)
+            };
+
+            if (isExactMatch) {
+              suggestion.exactMatch = true;
+            }
+
+            POSSIBLE_VARIATIONS.resources.push(suggestion);
+          } else if (isMinMatchingLength) {
+            const d = distance(normalizedStandardResourceName, normalizedResourceName),
+              maxDistance = Math.floor(fuzziness * resourceName?.length);
+
+            if (!hasStandardResource && d <= maxDistance) {
+              const suggestion = {
+                resourceName,
+                suggestedResourceName: standardResourceName,
+                distance: d,
+                maxDistance,
+                strategy: MATCHING_STRATEGIES.EDIT_DISTANCE,
+                ddWikiUrl: getDDWikiUrlForResourceName(standardResourceName)
+              };
+
+              if (d <= CLOSE_MATCH_DISTANCE) {
+                suggestion.closeMatch = true;
+              }
+
+              POSSIBLE_VARIATIONS.resources.push(suggestion);
+            }
+          }
+        });
+      } else {
+        //found standard resource - check field name variations
+        Object.keys(metadataReportMap?.[resourceName]).forEach(fieldName => {
```

Comment:
The pattern `Object.keys(metadataReportMap?.[resourceName])` would throw if `metadataReportMap?.[resourceName]` is undefined of null (`Object.keys(undefined | null)` throws). We could maybe use  `Object.keys(metadataReportMap?.[resourceName] || {})` ?

Reply:
Thanks. As a side note, have been using `??` rather than `||`.